File: typescript/examples/local_demo.ts
import * as os from 'os';
import * as path from 'path';
import { Keystore } from '../src/keystore';
import { FlashProvider } from '../src/wallet';
import { TronProvider } from '../src/wallet/tron_provider';

/**
 * Local-only demo that exercises SDK methods without network calls.
 *
 * Run:
 *   cd typescript
 *   npm run example:local-demo
 */

class LocalTronProvider extends TronProvider {
  /** Local stub: sign "transaction" without Tron RPC. */
  override async sign(transaction: any): Promise<any> {
    // For demo purposes, treat tx.txID as a hex string and sign that message.
    const txID = transaction?.txID ?? '00';
    const msg = typeof txID === 'string' ? txID : JSON.stringify(txID);
    const sig = await this.signMessage(Buffer.from(msg, 'utf8'));
    return { ...transaction, signature: [sig] };
  }

  override async getBalance(_address?: string): Promise<number> {
    return 123_456_789;
  }

  override async getTrc20Balance(_walletAddress: string, _contractAddress: string): Promise<string> {
    return '1000000';
  }

  override async sendTransaction(_toAddress: string, _amount: number): Promise<any> {
    return { result: true, txid: 'LOCAL_TXID' };
  }

  override async broadcast(_signedTransaction: any): Promise<any> {
    return { result: true, txid: 'LOCAL_BROADCAST_TXID' };
  }
}

class LocalFlashProvider extends FlashProvider {
  override async sign(transaction: any): Promise<any> {
    // Keep it local even if Privy creds exist.
    return (this as any).__proto__.__proto__.sign.call(this, transaction);
  }

  override async sendTransaction(_toAddress: string, _amount: number): Promise<any> {
    return { result: true, txid: 'LOCAL_FLASH_TXID' };
  }
}

async function main(): Promise<void> {
  const tmpKeystorePath = path.join(os.tmpdir(), `agent-wallet-keystore-${Date.now()}.bin`);

  // ===== Keystore (all methods) =====
  const ks = new Keystore({ filePath: tmpKeystorePath });
  console.log('[keystore] path:', ks.getPath());

  await ks.read();
  await ks.set('privateKey', '11'.repeat(32)); // demo-only key, do not use in production
  await ks.set('apiKey', 'demo-api-key');
  await ks.set('rpcUrl', 'http://localhost:9999');
  await ks.write();

  console.log('[keystore] keys:', await ks.keys());
  console.log('[keystore] privateKey len:', (await ks.get('privateKey'))?.length);
  console.log('[keystore] all:', await ks.getAll());

  const snap = await Keystore.fromFile(tmpKeystorePath);
  console.log('[keystore] fromFile:', snap);
  await Keystore.toFile(tmpKeystorePath, { ...snap, note: 'updated by toFile' });

  // ===== TronProvider (all methods, local stubs) =====
  const tron = new LocalTronProvider({ keystore: { filePath: tmpKeystorePath } });
  await tron.init();

  console.log('[tron] account:', await tron.getAccountInfo());
  console.log('[tron] signTx(message):', await tron.signTx({ type: 'message', message: Buffer.from('hello', 'utf8') }));
  console.log('[tron] signTx(tx):', await tron.signTx({ txID: 'tx-demo' }));
  console.log('[tron] sign(tx):', await tron.sign({ txID: 'tx-demo-2' }));
  console.log('[tron] balance:', await tron.getBalance());
  console.log('[tron] trc20 balance:', await tron.getTrc20Balance('wallet', 'contract'));
  console.log('[tron] sendTransaction:', await tron.sendTransaction('recipient', 1));
  console.log('[tron] broadcast:', await tron.broadcast({ signature: ['x'] }));

  // ===== FlashProvider (all methods, local stubs) =====
  const flash = new LocalFlashProvider({ keystore: { filePath: tmpKeystorePath } });
  await flash.init();
  console.log('[flash] signTx(message):', await flash.signTx({ type: 'message', message: Buffer.from('hello', 'utf8') }));
  console.log('[flash] sign(tx):', await flash.sign({ txID: 'flash-tx' }));
  console.log('[flash] sendTransaction:', await flash.sendTransaction('recipient', 2));
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: python/examples/local_demo.py
import asyncio
import os
import tempfile

import sys
from pathlib import Path

# Ensure `python/` directory is on sys.path when running as a script.
sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

from keystore import Keystore
from wallet import TronProvider, FlashProvider


class LocalTronProvider(TronProvider):
    """Local-only provider that overrides network methods."""

    async def get_balance(self, address=None) -> float:
        return 123.456

    async def get_trc20_balance(self, wallet_address: str, contract_address: str) -> int:
        return 1000000

    async def send_transaction(self, to_address: str, amount: float) -> dict:
        return {"result": True, "txid": "LOCAL_TXID"}

    async def broadcast_transaction(self, signed_transaction) -> dict:
        return {"result": True, "txid": "LOCAL_BROADCAST_TXID"}

    async def sign_transaction(self, transaction):
        # Local stub: attach a signature by signing a message derived from txid/txID
        txid = getattr(transaction, "txid", None) or getattr(transaction, "txID", None) or "tx-demo"
        sig = await self.sign_message(str(txid).encode("utf-8"))
        transaction._signature = [sig]
        return transaction


class LocalFlashProvider(FlashProvider):
    async def send_transaction(self, to_address: str, amount: float) -> dict:
        return {"result": True, "txid": "LOCAL_FLASH_TXID"}


async def main() -> int:
    tmp_path = os.path.join(tempfile.gettempdir(), f"agent-wallet-keystore-{os.getpid()}.bin")

    # ===== Keystore (all methods) =====
    ks = Keystore(file_path=tmp_path)
    print("[keystore] path:", ks.get_path())
    ks.read()
    ks.set("privateKey", "11" * 32)  # demo-only key, do not use in production
    ks.set("apiKey", "demo-api-key")
    ks.set("rpcUrl", "http://localhost:9999")
    ks.write()
    print("[keystore] keys:", sorted(ks.keys()))
    print("[keystore] privateKey len:", len(ks.get("privateKey") or ""))
    print("[keystore] all:", ks.get_all())
    snap = Keystore.from_file(tmp_path)
    print("[keystore] from_file:", snap)
    Keystore.to_file(tmp_path, {**snap, "note": "updated by to_file"})

    # ===== TronProvider (all methods, local stubs) =====
    tron = LocalTronProvider(keystore_path=tmp_path)
    await tron.init()
    print("[tron] account:", await tron.get_account_info())
    print("[tron] sign_tx(message):", await tron.sign_tx({"type": "message", "message": b"hello"}))
    # stub transaction object
    class Tx:
        txid = "tx-demo"
    signed_tx = await tron.sign_transaction(Tx())
    print("[tron] sign_transaction:", getattr(signed_tx, "_signature", None))
    print("[tron] balance:", await tron.get_balance())
    print("[tron] trc20 balance:", await tron.get_trc20_balance("wallet", "contract"))
    print("[tron] send_transaction:", await tron.send_transaction("recipient", 1.0))
    print("[tron] broadcast_transaction:", await tron.broadcast_transaction(signed_tx))

    # ===== FlashProvider (all methods, local stubs) =====
    flash = LocalFlashProvider(
        keystore_path=tmp_path,
        # keep privy creds unset so it falls back to local signing if used
        privy_app_id=None,
        privy_app_secret=None,
        wallet_id=None,
    )
    await flash.init()
    print("[flash] sign_tx(message):", await flash.sign_tx({"type": "message", "message": b"hello"}))
    print("[flash] send_transaction:", await flash.send_transaction("recipient", 2.0))

    return 0


if __name__ == "__main__":
    raise SystemExit(asyncio.run(main()))

